cpp/scenario/gazebo/src/Model.cpp

```diff
@@ -0,0 +1,1051 @@
+    };
+
+    return Impl::getJointDataSerialized(this, jointNames, lambda);
+}
+
+std::string Model::baseFrame() const
+{
+    // Get all the canonical links of the model
+    auto candidateBaseLinks = pImpl->ecm->EntitiesByComponents(
+        ignition::gazebo::components::CanonicalLink(),
+        ignition::gazebo::components::ParentEntity(pImpl->model.Entity()));
+
+    if (candidateBaseLinks.size() == 0) {
+        throw exceptions::ModelError("Failed to find the canonical link",
+                                     this->name());
+    }
+
+    if (candidateBaseLinks.size() > 1) {
+        throw exceptions::ModelError("Found multiple canonical link",
+                                     this->name());
+    }
+
+    // Get the name of the base link
+    std::string baseLinkName = utils::getExistingComponentData< //
+        ignition::gazebo::components::Name>(pImpl->ecm,
+                                            candidateBaseLinks.front());
+
+    return baseLinkName;
+}
+
+bool Model::setBaseFrame(const std::string& frameName)
+{
+    const auto linkNames = this->linkNames();
+
+    if (std::find(linkNames.begin(), linkNames.end(), frameName)
+        == linkNames.end()) {
+        gymppError
+            << "Failed to set the model base on the frame of nonexistent link '"
+            << frameName << "'" << std::endl;
+        return false;
+    }
+
+    if (frameName == this->baseFrame()) {
+        gymppDebug << "Frame '" << baseFrame()
+                   << "' is already the current model base" << std::endl;
+        return true;
+    }
+
+    gymppError << "Changing the base link is not yet supported" << std::endl;
+    throw exceptions::NotImplementedError(__FUNCTION__);
+}
+
+bool Model::fixedBase() const
+{
+    throw exceptions::NotImplementedError(__FUNCTION__);
+}
+
+bool Model::setAsFixedBase(const bool fixedBase)
+{
+    throw exceptions::NotImplementedError(__FUNCTION__);
+}
+
+std::array<double, 6> Model::baseWrench() const
+{
+    throw exceptions::NotImplementedError(__FUNCTION__);
+}
+
+std::array<double, 3> Model::basePosition() const
+{
+    // Get the model pose
+    ignition::math::Pose3d world_H_model = utils::getExistingComponentData< //
+        ignition::gazebo::components::Pose>(pImpl->ecm, pImpl->modelEntity);
+
+    return utils::fromIgnitionPose(world_H_model).position;
+}
+
+std::array<double, 4> Model::baseOrientation() const
+{
+    // Get the model pose
+    ignition::math::Pose3d world_H_model = utils::getExistingComponentData< //
+        ignition::gazebo::components::Pose>(pImpl->ecm, pImpl->modelEntity);
+
+    return utils::fromIgnitionPose(world_H_model).orientation;
+}
+
+std::array<double, 3> Model::baseLinearVelocity() const
+{
+    // We can get the velocity of the base link. Since there's only a rigid
+    // transformation between base and model frame, and the velocity is
+    // expressed in the world frame, we do not need to perform any conversion.
```

Comment:
I am not sure I follow. If we are discussing about the linear part of mixed velocity, if you just have a translation the linear velocity can change. Think for example of the mixed velocity of a disk rotating around a fixed axis: the mixed linear velocity of a frame centered in the axis is zero, while the one of a frame on the boundary of the disk is proportional to the angular velocity and the radius of the disk.  I think it would be definitely useful  if you could provide math details or link to the math details for the methods that actually manipulate velocity and accelerations.

Reply:
Sorry for not being clear enough. This comment is related to the discussion we had in https://github.com/robotology/gym-ignition/pull/117#discussion_r362244988. The outcome of that discussion was [`IgnitionRobot::baseVelocity()`](https://github.com/robotology/gym-ignition/blob/master/ignition/src/IgnitionRobot.cpp#L1241-L1264).

I am checking the math but now it seems to me that the linear part of the velocity could not be taken as it is but we have to apply a cross product considering the transform of the canonical link.

Reply:
Hopefully addressed in https://github.com/robotology/gym-ignition/pull/158/commits/ac70d11e0577da362814b0b8cd2d1643674992b6. Can you please double check the logic?

Reply:
Note: i'm going to move the helper methods to another file with another commit